# 2장 1강: t-검정의 이론과 가정 — 실습문제

## 실습 목표

- 평균 차이와 표준오차를 이용해 t통계량을 직접 계산할 수 있다.
- 단일표본 t검정으로 하나의 표본평균과 기준값을 비교할 수 있다.
- 독립성·정규성·등분산성을 확인한 뒤 독립표본 t검정을 수행할 수 있다.
- 독립표본과 대응표본 상황을 구분하고 적절한 함수를 선택할 수 있다.
- p-value를 유의수준과 비교하여 검정 결과를 올바르게 해석할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- `ames_housing.csv`

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `KitchenQual` | 주방 품질 |

> 모든 검정은 양측검정이며 유의수준 `α = 0.05`를 사용합니다.  
> Ames 표본 추출에는 지정된 `random_state`를 사용해 결과를 재현합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [2]:
# 실습 준비 코드를 작성하세요.
from pathlib import Path
import numpy as np
import pandas as pd
from scipy import stats
df = pd.read_csv("ames_housing.csv")

---

## 필수 1. t통계량 직접 계산과 단일표본 t검정

### 문제 1-1. 평균 판매가격은 180,000달러와 다른가?

#### 문제 설명

Ames 주택 30개를 표본으로 추출하여 평균 판매가격이 기준값 180,000달러와 다른지 확인합니다. 먼저 t통계량을 직접 계산한 뒤 `ttest_1samp()` 결과와 비교하세요.

#### 요구사항

1. `SalePrice`에서 `n=30`, `random_state=2`로 표본을 추출하여 `sale_sample`에 저장하세요.
2. 표본 수, 표본평균, 표본표준편차를 출력하세요.
3. `stats.sem()`을 이용해 표준오차를 계산하세요.
4. 다음 식으로 t통계량을 직접 계산하세요.  
   `t = (표본평균 - 기준값) / 표준오차`
5. Shapiro-Wilk 검정으로 표본의 정규성을 확인하세요.
6. 다음 가설을 작성하세요.
   - H₀: 모집단 평균 판매가격은 180,000달러이다.
   - H₁: 모집단 평균 판매가격은 180,000달러가 아니다.
7. `stats.ttest_1samp()`로 단일표본 t검정을 수행하세요.
8. 직접 계산한 t통계량과 함수가 반환한 t통계량을 비교하세요.
9. p-value를 이용해 귀무가설 기각 여부를 판단하세요.

#### 해석 질문

**Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
**Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
**Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?  
**Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?

#### 제출 결과

- 기술통계량과 정규성 결과
- 직접 계산한 표준오차와 t통계량
- 단일표본 t검정 결과
- 귀무가설 판단과 해석
- Q1~Q4 답변

In [3]:
# 필수 1 코드를 작성하세요.
alpha = 0.05
standard_price = 180000

sale_sample = df['SalePrice'].dropna().sample(n=30, random_state=2)

sample_mean = sale_sample.mean()
sample_std = sale_sample.std(ddof=1)
standard_error = stats.sem(sale_sample)
manual_t = (sample_mean - standard_price) / standard_error
shapiro_result = stats.shapiro(sale_sample)

# H0: 모집단 평균 판매가격은 180,000달러이다.
# h1: 모집단 평균 판매가격은 180,000달러가 아니다.
t_result = stats.ttest_1samp(sale_sample, popmean=standard_price)


# ---- 결과 출력 ----
print("표본수: ", len(sale_sample))
print("표본평균: ", sample_mean)
print("표본표준편차: ", sample_std)
print("표준오차: ", standard_error)
print("정규성 p-value: ", shapiro_result.pvalue)

print("\n직접 계산한 t통계량: ", manual_t)
print("함수가 계산한 t통계량: ", t_result.statistic)
print("t통계량 일치 여부: ", round(manual_t, 6) == round(t_result.statistic, 6))

print("\np-value: ", t_result.pvalue)

if t_result.pvalue < alpha:
    print(f"\n최종 판단: p-value({t_result.pvalue:.6f}) < 유의수준({alpha}) → 귀무가설 기각")
    print("→ 모집단 평균 판매가격은 180,000달러가 아니라고 볼 통계적 근거가 있습니다.")
else:
    print(f"\n최종 판단: p-value({t_result.pvalue:.6f}) >= 유의수준({alpha}) → 귀무가설 기각 실패")
    print("→ 모집단 평균 판매가격이 180,000달러와 다르다고 볼 통계적 근거가 부족합니다.")

표본수:  30
표본평균:  233323.8
표본표준편차:  86013.01167060874
표준오차:  15703.755576982507
정규성 p-value:  0.21330287894275524

직접 계산한 t통계량:  3.395608123075882
함수가 계산한 t통계량:  3.395608123075882
t통계량 일치 여부:  True

p-value:  0.002003295934828204

최종 판단: p-value(0.002003) < 유의수준(0.05) → 귀무가설 기각
→ 모집단 평균 판매가격은 180,000달러가 아니라고 볼 통계적 근거가 있습니다.


### 필수 1 답변 작성란

- **Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
<br> -> 관측된 평균과 기준값의 차이를 (그 차이의 불확실성을 읨하는 ) 표준오차로 나눈 값

- **Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
<br> -> 분모(표준오차)가 작아지므로 T통계량의 절대값은 커진다.

- **Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?  
<br> -> 일치한다. 두 방법 모두 약 3.3956으로 계산됨

- **Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?
<br> -> 그렇다. P-value가 약 0.002로 0.05보다 작아 귀무가설을 기각한다.
<br> -> 따라서 평균 판매가격은 180,000달러와 다르다

---

## 필수 2. 독립표본 t검정의 가정 점검과 수행

### 문제 2-1. 주방 품질 `Gd`와 `TA` 집단의 평균 판매가격 비교

#### 문제 설명

주방 품질이 `Gd`인 주택과 `TA`인 주택에서 각각 20개를 추출해 평균 판매가격을 비교합니다. 두 집단은 서로 다른 주택으로 구성되어 있습니다.

#### 요구사항

1. `KitchenQual == "Gd"`와 `KitchenQual == "TA"` 집단의 `SalePrice`에서 각각 `n=20`, `random_state=42`로 표본을 추출하세요.
2. 두 집단의 표본 수와 평균을 확인하세요.
3. 데이터 수집 구조를 근거로 두 집단의 독립성을 설명하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 두 집단에 Levene 등분산 검정을 수행하세요.
6. 가정 점검 결과에 따라 `equal_var=True` 또는 `equal_var=False`를 결정하세요.
7. `stats.ttest_ind()`로 독립표본 t검정을 수행하세요.
8. t통계량과 p-value를 출력하고 두 집단 평균 차이를 해석하세요.

#### 해석 질문

**Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
**Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
**Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
**Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
**Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 독립성 설명
- 정규성 및 등분산성 검정 결과
- `equal_var` 선택 근거
- 독립표본 t검정 결과와 해석
- Q1~Q5 답변

In [8]:
# 필수 2 코드를 작성하세요.

# 1. Gd, TA 집단에서 각각 20개 표본 추출
group_gd = (
    df.loc[df["KitchenQual"]=="Gd", "SalePrice"]
    .dropna()
    .sample(n=20, random_state=42)
)

group_ta = (
    df.loc[df["KitchenQual"]=="TA", "SalePrice"]
    .dropna()
    .sample(n=20, random_state=42)
)

# 2. 표본 수와 평균 확인
print("Gd 표본 수: ", len(group_gd))
print("TA 표본 수: ", len(group_ta))
print(f"Gd 평균: {group_gd.mean()}달러")
print(f"TA 평균: {group_ta.mean()}달러")

# 3. 독립성 설명 (주석)
# Gd 집단과 TA 집단은 서로 다른 주택(관측 단위)에서 추출되었고,
# 한 주택의 KitchenQual은 Gd 또는 TA 둘 중 하나로만 분류되므로
# 한 집단에 속한 관측치가 다른 집단에 속한 관측치와 겹치거나 서로 영향을 주지 않습니다.
# 즉, 같은 집을 반복 측정하거나 짝지은 게 아니라 "서로 다른 대상"을 비교하는 구조이므로
# 두 집단은 독립(independent)이라고 볼 수 있습니다.

# 4. 정규성 검정 (Shapiro-Wilk)
shapiro_gd = stats.shapiro(group_gd)
shapiro_ta = stats.shapiro(group_ta)

print("\n[Gd 정규성]")
print("statistic: ", shapiro_gd.statistic)
print("p-value: ", shapiro_gd.pvalue)


print("\n[TA 정규성]")
print("statistic: ", shapiro_ta.statistic)
print("p-value: ", shapiro_ta.pvalue)


# 5. 등분산성 검정 (Levene)
levene_result = stats.levene(group_gd, group_ta)
print("\n[등분산성]")
print("statistic: ", levene_result.statistic)
print("p-value: ", levene_result.pvalue)

alpha = 0.05
normal_gd = shapiro_gd.pvalue >= alpha
normal_ta = shapiro_ta.pvalue >= alpha
equal_var = levene_result.pvalue >= alpha

print(f"Gd 정규성 p-value: {shapiro_gd.pvalue: .4f} -> {'정규성 위반 근거 없음' if normal_gd else '정규성 위반'}")
print(f"TA 정규성 p-value: {shapiro_ta.pvalue: .4f} -> {'정규성 위반 근거 없음' if normal_ta else '정규성 위반'}")
print(f"등분산성 p-value: {levene_result.pvalue: .4f} -> {'등분산성 위반 근거 없음' if equal_var else '등분산성 위반'}")


# 6. 가정 점검 결과에 따라 equal_var 결정
alpha = 0.05
equal_var = levene_result.pvalue >= alpha
print("\nequal_var 설정: ", equal_var)

# 7. 독립표본 t검정 수행
t_result = stats.ttest_ind(group_gd, group_ta, equal_var=equal_var)

# 8. 결과 출력 및 해석
print("\nt통계량: ", t_result.statistic)
print("p-value: ", t_result.pvalue)

if t_result.pvalue < alpha:
    print(f"\np-value({t_result.pvalue:.6f}) < 유의수준({alpha})")
    print("→ 귀무가설 기각: Gd와 TA 집단의 평균 SalePrice는 통계적으로 유의미한 차이가 있습니다.")
else:
    print(f"\np-value({t_result.pvalue:.6f}) >= 유의수준({alpha})")
    print("→ 귀무가설 기각 실패: 두 집단의 평균 SalePrice가 다르다고 볼 통계적 근거가 부족합니다.")

Gd 표본 수:  20
TA 표본 수:  20
Gd 평균: 191916.6달러
TA 평균: 137432.5달러

[Gd 정규성]
statistic:  0.9554825623016634
p-value:  0.45798666908820035

[TA 정규성]
statistic:  0.9409863618579694
p-value:  0.2502820011486382

[등분산성]
statistic:  3.7805602341284543
p-value:  0.059281296941910915
Gd 정규성 p-value:  0.4580 -> 정규성 위반 근거 없음
TA 정규성 p-value:  0.2503 -> 정규성 위반 근거 없음
등분산성 p-value:  0.0593 -> 등분산성 위반 근거 없음

equal_var 설정:  True

t통계량:  3.424870993113913
p-value:  0.001489107014623552

p-value(0.001489) < 유의수준(0.05)
→ 귀무가설 기각: Gd와 TA 집단의 평균 SalePrice는 통계적으로 유의미한 차이가 있습니다.


### 필수 2 답변 작성란

- **Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
<br> -> GD/TA 각 집단이 서로 다른 주택으로 구성되어 하나의 관측값이 다른 관측값과 짝을 이루지 않기 때문이다.

- **Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
<br> -> 일반적으로 독립성은 별도의 검정 p-value로 확인하는 것이 아니라
<br> -> 데이터 수집 구조, Feature 엔지니어링 구조를 통해 판단하다.

- **Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
<br> -> 필수 2 문제에서는 정규성과 등분산성 위반 근거가 없다.

- **Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
<br> -> Levene p-value가 0.05보다 크므로 True를 사용한다.

- **Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?
<br> ->  유의한 차이가 있다, t 검정 결과 p-value가 약 0.0015로 0.05보다 작다.
<br> ->  일반적으로 주방설계를 잘해놓으면 주택가격이 달라질 가능성이 있다.

---

## 과제. 대응표본과 독립표본 상황 구분

### 문제 3-1. 같은 주택의 보수 전후 예상 판매가격 비교

#### 문제 설명

부동산 회사가 동일한 주택 10채에 대해 보수 전 예상 판매가격과 보수 후 예상 판매가격을 각각 산정했습니다. 같은 위치의 값은 동일한 주택의 전후 가격으로 서로 짝을 이룹니다.

```python
before_price = [145, 162, 178, 155, 190, 172, 168, 181, 159, 175]
after_price  = [154, 170, 185, 164, 201, 179, 176, 190, 168, 183]
```

단위는 천 달러입니다.

#### 요구사항

1. 두 배열의 길이가 같은지 확인하세요.
2. 보수 전후 평균을 계산하세요.
3. 독립표본 t검정과 대응표본 t검정 중 적절한 방법을 선택하고 이유를 설명하세요.
4. 대응표본 t검정의 정규성 가정은 개별 배열이 아니라 `after_price - before_price` 차이값에 적용된다는 점을 확인하세요.
5. 차이값에 Shapiro-Wilk 정규성 검정을 수행하세요.
6. 적절한 t검정을 실행하고 t통계량과 p-value를 출력하세요.
7. 보수 전후 평균 예상 판매가격에 유의한 차이가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  
**Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  
**Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  
**Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?

#### 제출 결과

- 표본 관계 판단과 근거
- 전후 평균과 차이값
- 차이값의 정규성 결과
- 대응표본 t검정 결과
- 결과 해석
- Q1~Q4 답변

In [ ]:
# 과제 코드를 작성하세요.

### 과제 답변 작성란

- **Q1.**
- **Q2.**
- **Q3.**
- **Q4.**

---

## 실습 마무리

1. t통계량은 어떤 두 값을 비교하여 계산하나요?
<br> -> 관측된 평균 차이를 표준오차(그 차이의 불확실성)와 비교한다.

2. 단일표본 t검정은 어떤 상황에서 사용하나요?
<br> -> 하나의 표본평균을 특정 기준값 또는 모집단  평균과 비교할 때 사용한다.

3. 독립표본과 대응표본은 데이터 수집 구조에서 어떤 차이가 있나요?
<br> -> 독립표본은 서로 다른 대상들로 구성되고, 대응표본은 같은 대상의 반복 측정값처럼 각 관측값이 짝을 이룬다.

4. 독립표본 t검정에서 정규성·등분산성·독립성은 각각 어떻게 확인하나요?
<br> -> Shapiro, Levene, 독립성은 데이터 수집 구조

5. 분산이 같다고 보기 어렵거나 확신할 수 없을 때 어떤 t검정을 사용할 수 있나요?
<br> -> equal_var = False인 welch t 검정을 사용할 수 있다.